### PR TITLE
Set up new method of using profiles

### PR DESCRIPTION
**Pull request by LorenDB**
Warning: this will effectively delete all profiles created with the old method. This is a major enhancement to #306.

### CODE REVIEW

**Comment by deepbluev7** [comment]
Cool stuff! I found a few small issues:

1. The diff is not very readable, because you rebased commits from master.
2. You are not changing the location of the Cache from what I can tell, which means stuff does get overwritten.
3. Do you really still need the NhekoSettings class?

File: src/MainWindow.cpp [hunk 1]
```diff
@@ -154,10 +153,15 @@ MainWindow::MainWindow(const QString profile, QWidget *parent)
         trayIcon_->setVisible(userSettings_->tray());
 
         if (hasActiveUser()) {
-                QString token       = settings.value("auth/access_token").toString();
-                QString home_server = settings.value("auth/home_server").toString();
-                QString user_id     = settings.value("auth/user_id").toString();
-                QString device_id   = settings.value("auth/device_id").toString();
+                UserSettings us;
+                QString prefix;
+                if (us.profile() != "")
```
Comment (deepbluev7) [inline src/MainWindow.cpp:158]: Why is accessing userSettings_ not enough?

File: src/MainWindow.cpp [hunk 2]
```diff
@@ -337,8 +347,14 @@ MainWindow::hasActiveUser()
 {
         QSettings settings;
 
-        return settings.contains("auth/access_token") && settings.contains("auth/home_server") &&
-               settings.contains("auth/user_id");
+        UserSettings us;
```
Comment (deepbluev7) [inline src/MainWindow.cpp:350]: Why read the settings twice?

Comment (LorenDB) [reply src/MainWindow.cpp:350]: Please elaborate.

File: src/MainWindow.cpp (hunk 1 above)
Comment (LorenDB) [reply src/MainWindow.cpp:158]: That's a good point.

File: src/MainWindow.cpp (hunk 2 above)
Comment (deepbluev7) [reply src/MainWindow.cpp:350]: UserSettings uses QSettings. I don't see, why you would need to parse the same settings file twice. UserSettings should just provide access_token, homeserver, etc

Comment (LorenDB) [reply src/MainWindow.cpp:350]: Gotcha.

File: src/Cache.cpp [hunk 3]
```diff
@@ -21,10 +21,10 @@
 
 #include <QByteArray>
 #include <QCoreApplication>
+#include <QDebug>
```
Comment (deepbluev7) [inline src/Cache.cpp:24]: You should not need this.

File: src/MainWindow.cpp [hunk 4]
```diff
@@ -154,10 +155,14 @@ MainWindow::MainWindow(const QString profile, QWidget *parent)
...
-                QString user_id     = settings.value("auth/user_id").toString();
-                QString device_id   = settings.value("auth/device_id").toString();
+                QString prefix;
+                if (userSettings_->profile() != "")
+                        prefix = "profile/" + userSettings_->profile() + "/";
+
+                QString token       = settings.value(prefix + "auth/access_token").toString();
+                QString home_server = settings.value(prefix + "auth/home_server").toString();
+                QString user_id     = settings.value(prefix + "auth/user_id").toString();
+                QString device_id   = settings.value(prefix + "auth/device_id").toString();
```
Comment (deepbluev7) [inline src/MainWindow.cpp:165]: Can you do that the same way as you do it in showChatPage?

File: src/main.cpp [hunk 5]
```diff
@@ -185,16 +161,37 @@ main(int argc, char *argv[])
         if (parser.isSet(debugOption))
                 nhlog::enable_debug_log_from_commandline = true;
 
+        QString appNameWithProfile{"nheko"};
+        if (parser.isSet(configName))
+                appNameWithProfile.append(QString{"-%1"}.arg(parser.value(configName)));
+
+        QSharedMemory mem{appNameWithProfile};
```
Comment (deepbluev7) [inline src/main.cpp:168]: Do we still need SingleApplication then?

File: src/main.cpp [hunk 6]
```diff
@@ -185,16 +161,37 @@ main(int argc, char *argv[])
...
+                nhlog::init(QString("%1/%2.log")
                               .arg(QStandardPaths::writableLocation(QStandardPaths::CacheLocation))
+                              .arg(appNameWithProfile)
                               .toStdString());
         } catch (const spdlog::spdlog_ex &ex) {
                 std::cout << "Log initialization failed: " << ex.what() << std::endl;
                 std::exit(1);
         }
 
         QSettings settings;
```
Comment (deepbluev7) [inline src/main.cpp:190]: Do we still need this, if we have UserSettings below?

File: src/main.cpp (hunk 5 above)
Comment (LorenDB) [reply src/main.cpp:168]: Probably not.

File: src/Cache.cpp (hunk 3 above)
Comment (LorenDB) [reply src/Cache.cpp:24]: Oops

File: src/main.cpp (hunk 6 above)
Comment (LorenDB) [reply src/main.cpp:190]: Please elaborate.

File: src/MainWindow.cpp (hunk 2 above)
Comment (LorenDB) [reply src/MainWindow.cpp]: Could the part that goes:
```
        return settings.contains("auth/access_token") && settings.contains("auth/home_server") &&
               settings.contains("auth/user_id");
```
be switched to something like:
```
        return userSettings_->userId() != "" && ...
```

**Comment by deepbluev7** [comment]
One thing that is broken by this:

Nheko used to activate the current window, if one started the same profile again. Can you add that again? It depended on SingleApplication, but your QSharedMemory stuff broke it and never let it get that far.

**Comment by LorenDB** [comment]
> One thing that is broken by this:
> 
> Nheko used to activate the current window, if one started the same profile again. Can you add that again? It depended on SingleApplication, but your QSharedMemory stuff broke it and never let it get that far.

It would be nice to bring that back; however, to do so will require one of two options:

- Not allow multiple active instances.
- Ditch QSharedMemory in favor of using either Qt's dbus service or QTcpSocket/Server to allow triggering a raise of the current window when necessary.

Any comments would be appreciated.

**Comment by deepbluev7** [comment]
> It would be nice to bring that back; however, to do so will require one of two options:
> 
>     * Not allow multiple active instances.
> 
>     * Ditch QSharedMemory in favor of using either Qt's dbus service or QTcpSocket/Server to allow triggering a raise of the current window when necessary.
> 
> 
> Any comments would be appreciated.

Well, SingleApplication does implement exactly that, but it is interfering with your profile stuff. The right way to fix that, is probably to add additional "userdata" one can specify to SingleApplication, which can be used to distinguish applications. That would solve your issue as well and we wouldn't need to support the QSharedMemory stuff ourselves.

File: src/main.cpp (hunk 6 above)
Comment (deepbluev7) [reply src/main.cpp:214]: the settings here still looks redundant, if you have UserSettings right below it

File: src/main.cpp [hunk 7]
```diff
@@ -185,16 +185,37 @@ main(int argc, char *argv[])
...
+        // Try to create the shared memory. If it already exists, we'll exit
+        if (!mem.create(1)) {
+                // attach and detach to try to delete hanging shared memory
+                // this would be caused by a crash on Unix-based systems (e.g. on a segfault)
+                mem.attach();
+                mem.detach();
+                if (!mem.create(1))
+                        exit(0);
+        }
+
```
Comment (deepbluev7) [inline src/main.cpp:203]: This can be removed now?

File: src/main.cpp (hunk 6 above)
Comment (LorenDB) [reply src/main.cpp:214]: Hmm, I think that the QSettings is still used in main(), so I'll leave it for now.

File: src/main.cpp [hunk 8]
```diff
@@ -195,6 +197,10 @@ main(int argc, char *argv[])
         }
 
         QSettings settings;
```
Comment (deepbluev7) [inline src/main.cpp:199]: You can still delete this!

```suggestion
```

File: src/Cache.cpp [hunk 9]
```diff
@@ -210,7 +214,6 @@ Cache::setup()
                                 throw std::runtime_error(
                                   ("Unable to delete file " + file).toStdString().c_str());
                 }
-
                 env_.open(statePath.toStdString().c_str());
         }
```
Comment (deepbluev7) [inline src/Cache.cpp:219]: You need to now also patch the storeSecret, secret and removeSecret functions.

File: src/main.cpp (hunk 8 above)
Comment (LorenDB) [reply src/main.cpp:199]: Hmm, looks like somebody added some things to UserSettings but never used them in main...

Comment (LorenDB) [reply src/main.cpp:199]: Actually, there are things in UserSettings to set font size & family and the start in tray values (what the QSettings is used for in main()), but not to get them... Looks like I've got a bit more work :P

Comment (deepbluev7) [reply src/main.cpp:199]: Well, I just don't like reading the settings file twice immediately in main :D

Comment (LorenDB) [reply src/main.cpp:199]: OK, forget my previous comment

File: src/Cache.cpp (hunk 9 above)
Comment (LorenDB) [reply src/Cache.cpp:219]: I'll get to that on Monday.

Comment (deepbluev7) [reply src/Cache.cpp:219]: Sure, take your time :3

File: src/Cache.cpp [hunk 10]
```diff
@@ -175,6 +177,9 @@ Cache::setup()
                             .arg(QStandardPaths::writableLocation(QStandardPaths::CacheLocation))
                             .arg(QString::fromUtf8(localUserId_.toUtf8().toHex()));
 
+        UserSettings settings;
+        statePath += settings.profile();
```
Comment (deepbluev7) [inline src/Cache.cpp:181]: Please make statePath and cacheDirectory_ the same. Currently this makes this delete the wrong directory on Logout. Even better, if you just remove the statePath.

File: src/Cache.cpp [hunk 11]
```diff
@@ -580,7 +583,10 @@ Cache::storeSecret(const std::string &name, const std::string &secret)
         QKeychain::WritePasswordJob job(QCoreApplication::applicationName());
         job.setAutoDelete(false);
         job.setInsecureFallback(true);
-        job.setKey(QString::fromStdString(name));
+        job.setKey(
+          "matrix." +
+          QString(QCryptographicHash::hash(localUserId_.toUtf8(), QCryptographicHash::Sha256)) +
```
Comment (deepbluev7) [inline src/Cache.cpp:588]: Please use the profile here, not the userid. Otherwise logout deletes the wrong secrets.

File: src/MainWindow.cpp [hunk 12]
```diff
@@ -55,10 +55,12 @@
 
 MainWindow *MainWindow::instance_ = nullptr;
 
-MainWindow::MainWindow(const QString profile, QWidget *parent)
+MainWindow::MainWindow(QWidget *parent)
   : QMainWindow(parent)
-  , profile_{profile}
 {
+        // initialize this here so that setWindowTitle can use it
+        userSettings_ = QSharedPointer<UserSettings>(new UserSettings);
```
Comment (deepbluev7) [inline src/MainWindow.cpp:62]: Just move this to the initializer list at this point :3

**Review by deepbluev7 — COMMENTED** [review]
Cheers. I finally found time to try this out, but I found some bugs while doing that. :D

**Comment by deepbluev7** [comment]
:tada:

**Comment by LorenDB** [comment]
Yeah!!! Finally!

**Pull request merged**